src/transformers/models/bloom/modeling_bloom.py

```diff
@@ -278,54 +277,67 @@ def forward(
         # 3 x [batch_size, seq_length, num_heads, head_dim]
         (query_layer, key_layer, value_layer) = self._split_heads(fused_qkv)
 
+        batch_size, q_length, _, _ = query_layer.shape
+
+        query_layer = query_layer.transpose(1, 2).reshape(batch_size * self.num_heads, q_length, self.head_dim)
+        key_layer = key_layer.permute(0, 2, 3, 1).reshape(batch_size * self.num_heads, self.head_dim, q_length)
+        value_layer = value_layer.transpose(1, 2).reshape(batch_size * self.num_heads, q_length, self.head_dim)
         if layer_past is not None:
             past_key, past_value = layer_past
-            # concatenate along seq_length dimension -> [batch_size, qk_length, num_heads, head_dim]
-            key_layer = torch.cat((past_key.type_as(key_layer), key_layer), dim=1)
+            # concatenate along seq_length dimension:
+            #  - key: [batch_size * self.num_heads, head_dim, kv_length]
+            #  - value: [batch_size * self.num_heads, kv_length, head_dim]
+            key_layer = torch.cat((past_key.type_as(key_layer), key_layer), dim=2)
             value_layer = torch.cat((past_value.type_as(value_layer), value_layer), dim=1)
```

Comment:
This prevents a bunch of unecessary copies downstream.

Reply:
Hum actually this is probably a breaking change, if we expose `past_key_values`.

Reply:
Tried having a non breaking change (though there would have still been the contiguougs issue): https://github.com/huggingface/transformers/pull/18344/files/5ed059c1d7bf29e48165e642590a228fce4f4fda..4596be9e57ac8e2e5e0ddfd0a7ed4f48e32c8c4a however this fails because you can't run `transpose().view()`